File: brkga_mp_ipr_java/bench/java/TspDecoderOptimized.java
/******************************************************************************
 * TspDecoderOptimized.java: optimized single-objective TSP decoder.
 *
 * Same RESULT as bench.TspDecoder, but tuned for the hot path with the very same
 * technique as the C++ bench TSPDecoderOptimized, so the two stay equivalent:
 *
 *   1. A reusable per-thread buffer (ThreadLocal long[]), so no allocation per
 *      decode.
 *   2. A primitive packed sort: each gene's IEEE-754 bit pattern (monotonic for
 *      the non-negative keys in [0,1)) is packed in the high bits and the node
 *      index in the low `indexBits` bits, so a single ascending Arrays.sort(long[])
 *      orders by gene and breaks ties by index — no boxing, no comparator.
 *****************************************************************************/

package bench;

import brkga.Chromosome;
import brkga.Decoder;

import java.util.Arrays;

public final class TspDecoderOptimized implements Decoder {

    public final TspInstance instance;

    private final long indexMask;
    private final long keyMask;
    private final ThreadLocal<long[]> packedPerThread;

    public TspDecoderOptimized(TspInstance instance) {
        this.instance = instance;
        final int n = instance.numNodes;
        int indexBits = Math.max(1, 32 - Integer.numberOfLeadingZeros(Math.max(1, n - 1)));
        this.indexMask = (1L << indexBits) - 1L;
        this.keyMask = ~indexMask;
        this.packedPerThread = ThreadLocal.withInitial(() -> new long[n]);
    }

    @Override
    public double[] decode(Chromosome chromosome, boolean rewrite) {
        final int size = chromosome.size();
        final long[] packed = packedPerThread.get();

        for(int i = 0; i < size; ++i) {
            long bits = Double.doubleToRawLongBits(chromosome.get(i));
            packed[i] = (bits & keyMask) | i;
        }

        Arrays.sort(packed, 0, size);

        int first = (int) (packed[0] & indexMask);
        int last = (int) (packed[size - 1] & indexMask);
        double cost = instance.distance(first, last);

        for(int i = 0; i < size - 1; ++i)
            cost += instance.distance((int) (packed[i] & indexMask),
                                      (int) (packed[i + 1] & indexMask));

        return new double[] { cost };
    }
}

File: brkga_mp_ipr_java/bench/cpp/tsp_decoder_optimized.cpp
/******************************************************************************
 * tsp_decoder_optimized.cpp: implementation of the optimized TSP decoder.
 *****************************************************************************/

#include "tsp_decoder_optimized.hpp"

#include <algorithm>
#include <bit>
#include <vector>

TSPDecoderOptimized::TSPDecoderOptimized(const TSPInstance& inst):
    instance(inst)
{
    // Bits needed to hold a node index in [0, num_nodes). Matches the Java side:
    //   indexBits = max(1, 32 - numberOfLeadingZeros(max(1, n - 1))).
    unsigned n = inst.num_nodes;
    unsigned index_bits = 1;
    while((1u << index_bits) < n)
        ++index_bits;
    index_mask = (std::uint64_t(1) << index_bits) - 1;
    key_mask = ~index_mask;
}

BRKGA::fitness_t TSPDecoderOptimized::decode(BRKGA::Chromosome& chromosome,
                                             bool /* rewrite, unused */) {
    const std::size_t n = chromosome.size();

    // Reused across calls within each (OpenMP) thread: no per-decode allocation.
    static thread_local std::vector<std::uint64_t> packed;
    if(packed.size() < n)
        packed.resize(n);

    for(std::size_t i = 0; i < n; ++i) {
        std::uint64_t bits = std::bit_cast<std::uint64_t>(chromosome[i]);
        packed[i] = (bits & key_mask) | i;
    }

    std::sort(packed.begin(), packed.begin() + n);

    unsigned first = static_cast<unsigned>(packed[0] & index_mask);
    unsigned last = static_cast<unsigned>(packed[n - 1] & index_mask);
    double cost = instance.distance(first, last);

    for(std::size_t i = 0; i < n - 1; ++i)
        cost += instance.distance(
            static_cast<unsigned>(packed[i] & index_mask),
            static_cast<unsigned>(packed[i + 1] & index_mask));

    return cost;
}
